[WebXR] Port enum classes to new serialization format

csaavedra opened the pull request:
#### 3fb2a422f959942a00e037541091558e8c3b18ed
<pre>
[WebXR] Port enum classes to new serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=265665">https://bugs.webkit.org/show_bug.cgi?id=265665</a>

Reviewed by Dan Glastonbury.

Move the XR enums to the new serialization format to be able to get
rid of the enum traits for them.

 * Source/WebCore/platform/xr/PlatformXR.h:
 * Source/WebKit/Shared/XR/XRSystem.serialization.in:

Canonical link: <a href="https://commits.webkit.org/271466@main">https://commits.webkit.org/271466@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64634841c2b1c001176e17dd2ca5589452c5b573

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28285 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6929 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29656 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30814 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25766 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28782 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8980 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4305 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28554 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5697 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24341 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4967 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5086 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25339 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31502 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25900 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25773 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31388 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5057 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3238 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29143 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6643 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6825 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5499 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5575 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebKit/Shared/XR/XRSystem.serialization.in:
```diff
@@ -32,6 +32,39 @@ enum class PlatformXR::SessionFeature : uint8_t {
+    ImmersiveAr,
+};
+
+[Nested] enum class PlatformXR::ReferenceSpaceType : uint8_t {
```
Q: For my own understanding, why does this one need the `[Nested]` attribute?

djg replied:
I compiled without `[Nested]` and it works, so I'd suggest removing it.

csaavedra replied:
My understanding is that you can use Nested if the class is only serialized as part of another class, simplifying a bit the declaration. Removing `[Nested]` adds the following forward declarations to the `GeneratedSerializers.h` header, which is not really necessary in this case:

```diff
@@ -73,6 +73,9 @@ enum class SessionFeature : uint8_t;
 enum class SessionMode : uint8_t;
 #endif
 #if ENABLE(WEBXR)
+enum class ReferenceSpaceType : uint8_t;
+#endif
+#if ENABLE(WEBXR)
 enum class VisibilityState : uint8_t;
 #endif
 #if ENABLE(WEBXR)
@@ -7688,6 +7691,9 @@ template<> bool isValidEnum<PlatformXR::
 template<> bool isValidEnum<PlatformXR::SessionMode, void>(uint8_t);
 #endif
 #if ENABLE(WEBXR)
+template<> bool isValidEnum<PlatformXR::ReferenceSpaceType, void>(uint8_t);
+#endif
+#if ENABLE(WEBXR)
 template<> bool isValidEnum<PlatformXR::VisibilityState, void>(uint8_t);
 #endif
 #if ENABLE(WEBXR)
```

csaavedra replied:
@djg Based on the explanation above I am going to merge-as-is, please let me know if you disagree.

djg reviewed (approved):
:shipit:

djg reviewed (approved):
:shipit:

csaavedra committed:
[WebXR] Port enum classes to new serialization format
https://bugs.webkit.org/show_bug.cgi?id=265665

Reviewed by Dan Glastonbury.

Move the XR enums to the new serialization format to be able to get
rid of the enum traits for them.

 * Source/WebCore/platform/xr/PlatformXR.h:
 * Source/WebKit/Shared/XR/XRSystem.serialization.in:

Canonical link: https://commits.webkit.org/271466@main
